samza-core/src/main/java/org/apache/samza/task/TaskFactoryUtil.java

```diff
@@ -0,0 +1,156 @@
+      throw new ConfigException(String.format("Invalid configuration for AsyncStreamTask class: %s", taskClassName), t);
+    }
+
+    if (isAsyncTaskClass) {
+      return new AsyncStreamTaskFactory() {
+        @Override
+        public AsyncStreamTask createInstance() {
+          try {
+            return (AsyncStreamTask) Class.forName(taskClassName).newInstance();
+          } catch (Throwable t) {
+            log.error("Error loading AsyncStreamTask class: {}. error: {}", taskClassName, t);
+            throw new SamzaException(String.format("Error loading AsyncStreamTask class: %s", taskClassName), t);
+          }
+        }
+      };
+    }
+
+    return new StreamTaskFactory() {
+      @Override
+      public StreamTask createInstance() {
+        try {
+          if (taskClassName.equals(StreamOperatorTask.class.getName())) {
+            return createStreamOperatorTask(config, runner);
+          }
+          return (StreamTask) Class.forName(taskClassName).newInstance();
+        } catch (Throwable t) {
+          log.error("Error loading StreamTask class: {}. error: {}", taskClassName, t);
+          throw new SamzaException(String.format("Error loading StreamTask class: %s", taskClassName), t);
+        }
+      }
+    };
+  }
+
+  public static TaskFactory finalizeTaskFactory(TaskFactory factory, boolean singleThreadMode, ExecutorService taskThreadPool) {
+
+    validateFactory(factory);
+
+    boolean isAsyncTaskClass = factory instanceof AsyncStreamTaskFactory;
+    if (isAsyncTaskClass) {
+      log.info("Got an AsyncStreamTask implementation.");
+    }
+
+    if (singleThreadMode && isAsyncTaskClass) {
+      throw new SamzaException("AsyncStreamTask cannot run on single thread mode.");
+    }
+
+    if (!singleThreadMode && !isAsyncTaskClass) {
+      log.info("Converting StreamTask to AsyncStreamTaskAdapter when running StreamTask with multiple threads");
+      return new AsyncStreamTaskFactory() {
+        @Override
+        public AsyncStreamTask createInstance() {
+          return new AsyncStreamTaskAdapter((StreamTask) factory.createInstance(), taskThreadPool);
+        }
+      };
+    }
+
+    return factory;
+  }
+
+  private static void validateFactory(TaskFactory factory) {
+    if (factory == null) {
+      throw new SamzaException("Either the task class name or the task factory instance is required.");
+    }
+
+    if (!(factory instanceof StreamTaskFactory) && !(factory instanceof AsyncStreamTaskFactory)) {
+      throw new SamzaException(String.format("TaskFactory must be either StreamTaskFactory or AsyncStreamTaskFactory. %s is not supported",
+          factory.getClass()));
+    }
+  }
+
+  private static StreamTask createStreamOperatorTask(Config config, ApplicationRunner runner) throws Exception {
+    Class<?> builderClass = Class.forName(config.get(StreamApplication.APP_CLASS_CONFIG));
+    StreamApplication graphBuilder = (StreamApplication) builderClass.newInstance();
```

Comment:
This would be the same for a lot of classes we load. We would either need to change them all, or just follow the current pattern. Besides, the constructor w/o input parameter seems to be OK to me.